econml/sklearn_extensions/linear_model.py

```diff
@@ -1682,44 +1682,57 @@ def __init__(self, fit_intercept=True, cov_type=None):
-        if y.ndim >= 2:
-            assert (y.ndim == sample_var.ndim and
-                    y.shape[1] == sample_var.shape[1]), "Input shapes not compatible: {}, {}!".format(
-                y.shape, sample_var.shape)
+        if sample_weight is None:
+            sample_weight = np.ones(y.shape[0])
 
-        return X, y, sample_weight, sample_var
+        if sample_var is not None:
+            if np.any(np.not_equal(np.mod(sample_weight, 1), 0)):
+                raise AttributeError("Sample weights must all be integers for inference to be valid!")
+
+            if sample_var.ndim < 2:
+                if np.any(np.equal(sample_weight, 1) & np.not_equal(sample_var, 0)):
+                    warnings.warn(
+                        "Variance was set to non-zero for an observation with sample_weight=1! "
+                        "sample_var represents the variance of the original observations that are "
+                        "summarized in this sample. Hence, cannot have a non-zero variance if only "
+                        "one observations was summarized. Inference will be invalid!")
+            else:
+                if np.any(np.equal(sample_weight, 1) & np.not_equal(np.sum(sample_var, axis=1), 0)):
+                    warnings.warn(
+                        "Variance was set to non-zero for an observation with sample_weight=1! "
+                        "sample_var represents the variance of the original observations that are "
+                        "summarized in this sample. Hence, cannot have a non-zero variance if only "
+                        "one observations was summarized. Inference will be invalid!")
+            assert (X.shape[0] == y.shape[0] ==
+                    sample_weight.shape[0] == sample_var.shape[0]), "Input lengths not compatible!"
+            if y.ndim >= 2:
+                assert (y.ndim == sample_var.ndim and
+                        y.shape[1] == sample_var.shape[1]), "Input shapes not compatible: {}, {}!".format(
+                    y.shape, sample_var.shape)
+            return X, y, sample_weight, sample_var
+
+        else:
+            sample_var = np.zeros(y.shape)
+            assert (X.shape[0] == y.shape[0] == sample_weight.shape[0] ==
+                    sample_var.shape[0]), "Input lengths not compatible!"
+            if y.ndim >= 2:
+                assert (y.ndim == sample_var.ndim and
+                        y.shape[1] == sample_var.shape[1]), "Input shapes not compatible: {}, {}!".format(
+                    y.shape, sample_var.shape)
+
+            weighted_X = X * np.sqrt(sample_weight).reshape(-1, 1)
+            if y.ndim < 2:
+                weighted_y = y * np.sqrt(sample_weight)
+            else:
+                weighted_y = y * np.sqrt(sample_weight).reshape(-1, 1)
+            new_weight = np.ones_like(sample_weight)
+            return weighted_X, weighted_y, new_weight, sample_var
```

Comment:
Realized this reduction is ok for the HC cov_types, but it s not correct for the nonrobust, where you want to take var_i to be the unweighted residuals and have w_i be inside Sigma.

Reply:
Why? `param` is learnt by fit WY on WX, why calculate the residual using X and Y? And I compare with WLS `cov_params`, currently implementation matches exactly with WLS on different cov_type, but doesn't match if I use residuals as unweighted for nonrobust.

Reply:
Hm. surprised that it matches. Can you try weights that depend on the features? I think in that case the non-robust wont match. Basically the non-robust does the following: it takes the population quantity:
E[ w^2 (y - X'theta)^2 X X^T ] = E[(y - X'theta)^2] * E[w^2 XX^T]
where this can be achieved by assuming that E[(Y-X'\theta)^2|X] = Var(Epsilon) is independent of X.

But the current computation does:
E[ w^2 (y - X'theta)^2 X X^T ] = E[w^2 * (y - X'theta)^2] * E[XX^T]

which implicitly assumes that E[w^2|X] is also independent of X, which many times might not be the case.


Alsp I think we should make the default cov_type to be HC0, not None!